beacon-chain/chaintest/backend/helpers.go

```diff
@@ -0,0 +1,135 @@
+	numEpochs := numSlots % params.BeaconConfig().EpochLength
+	hashOnions := [][32]byte{params.BeaconConfig().SimulatedBlockRandao}
+
+	// We make the length of the hash onions list equal to the number of epochs + 10 to be safe.
+	for i := uint64(0); i < numEpochs+10; i++ {
+		prevHash := hashOnions[i]
+		hashOnions = append(hashOnions, hashutil.Hash(prevHash[:]))
+	}
+	return hashOnions
+}
+
+// This function determines the block randao reveal assuming there are no skipped slots,
+// given a list of randao hash onions such as [pre-image, 0x01, 0x02, 0x03], for the
+// 0th epoch, the block randao reveal will be 0x02 and the proposer commitment 0x03.
+// The next epoch, the block randao reveal will be 0x01 and the commitment 0x02,
+// so on and so forth until all randao layers are peeled off.
+func determineSimulatedBlockRandaoReveal(layersPeeled int, hashOnions [][32]byte) [32]byte {
+	if layersPeeled == 0 {
+		return hashOnions[len(hashOnions)-2]
+	}
+	return hashOnions[len(hashOnions)-layersPeeled-2]
+}
+
+// Generates initial deposits for creating a beacon state in the simulated
+// backend based on the yaml configuration.
+func generateInitialSimulatedDeposits(randaoCommit [32]byte) ([]*pb.Deposit, error) {
+	genesisTime := params.BeaconConfig().GenesisTime.Unix()
+	deposits := make([]*pb.Deposit, params.BeaconConfig().DepositsForChainStart)
+	for i := 0; i < len(deposits); i++ {
+		depositInput := &pb.DepositInput{
+			Pubkey:                 []byte(strconv.Itoa(i)),
+			RandaoCommitmentHash32: randaoCommit[:],
+		}
+		depositData, err := b.EncodeDepositData(
+			depositInput,
+			params.BeaconConfig().MaxDepositInGwei,
+			genesisTime,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("could not encode initial block deposits: %v", err)
+		}
+		deposits[i] = &pb.Deposit{DepositData: depositData}
+	}
+	return deposits, nil
+}
+
+// Finds the index of the next slot's proposer in the beacon state's
+// validator set.
+func findNextSlotProposerIndex(beaconState *pb.BeaconState) (uint32, error) {
+	nextSlot := beaconState.Slot + 1
+	epochLength := params.BeaconConfig().EpochLength
+	var earliestSlot uint64
+
+	// If the state slot is less than epochLength, then the earliestSlot would
+	// result in a negative number. Therefore we should default to
+	// earliestSlot = 0 in this case.
+	if nextSlot > epochLength {
+		earliestSlot = nextSlot - (nextSlot % epochLength) - epochLength
+	}
+
+	if nextSlot < earliestSlot || nextSlot >= earliestSlot+(epochLength*2) {
+		return 0, fmt.Errorf("slot %d out of bounds: %d <= slot < %d",
+			nextSlot,
+			earliestSlot,
+			earliestSlot+(epochLength*2),
+		)
+	}
+	committeeArray := beaconState.ShardAndCommitteesAtSlots[nextSlot-earliestSlot]
+	firstCommittee := committeeArray.ArrayShardAndCommittee[0].Committee
+	return firstCommittee[nextSlot%uint64(len(firstCommittee))], nil
+}
+
+// determines if a uint64 item exists in a slice.
+func contains(item uint64, slice []uint64) bool {
```

Comment:
we could place this in the shared slice helpers

Reply:
Good point, will do